System/Random/Internal.hs

```diff
@@ -688,20 +681,66 @@ foreign import prim "stg_word64ToDoubleyg"
     stgWord64ToDouble :: Word64# -> Double#
 #endif
 
-
 instance UniformRange Float where
   uniformRM (l, h) g = do
-    w32 <- uniformWord32 g
-    let x = word32ToFloatInUnitInterval w32
+    x <- floatInUnitIntervalM g
     return $ (h - l) * x + l
 
--- | Turns a given uniformly distributed 'Word32' value into a uniformly
--- distributed 'Float' value in the range [0,1).
-word32ToFloatInUnitInterval :: Word32 -> Float
-word32ToFloatInUnitInterval w32 = between1and2 - 1.0
+-- | Counts the number of failed Bernoulli trials with p=0.5 before the first
+-- success, up to the given limit.
+geometricDistr :: MonadRandom g s m => Int -> Int -> g s -> m Int
+geometricDistr start limit g = go start
   where
-    between1and2 = castWord32ToFloat $ (w32 `unsafeShiftR` 9) .|. 0x3f800000
-{-# INLINE word32ToFloatInUnitInterval #-}
+    go !acc
+      | acc >= limit = return limit
+      | otherwise = do
+          w <- uniformWord64 g
+          if w /= 0
+            then return $ acc + countLeadingZeros w
+            else go (acc + finiteBitSize w)
+{-# INLINE geometricDistr #-}
+
+-- | Generates a 'Float' in [0,1].
+floatInUnitIntervalM :: MonadRandom g s m => g s -> m Float
+floatInUnitIntervalM g = do
+  let maxExp = 126 -- exponent of 1.0, decremented by one
+  let mantissaBits = 23
+  let bernoulliBits = 64 - mantissaBits - 1
+  let carryMask = 2 ^ mantissaBits
+  let mantissaMask = (2 ^ mantissaBits) - 1
+
+  w <- uniformWord64 g
```

Comment:
Should it be `uniformWord32` for `Float`?

Reply:
We're using all the bits: everything that is not used for the mantissa or the carry bit is used for the exponent (see line 716).

We _could_ work with Word32 here. This works out as follows:

- if we generate a Word32 here then there are 8 remaining bits for the Bernoulli trials to generate the exponent. In other words, we have to call into the PRNG again (via `geometricDistr`) with a probability of 2^-8.
- if we generate a Word64 here then 40 bits remain for the Bernoulli trials, so `geometricDistr` only needs to be called with a probability of 2^-40.

So that's the trade-off. The assumption here is that it is better to generate a Word64 and fall back to `geometricDistr` with p=2^-40 than it is to generate a Word32 and fall back to `geometricDistr` with p=2^-8.

That's something that is easy to benchmark though.

Reply:
This makes sense. It is better to stick to `uniformWord64`, since most of good quality RNGs have to generate `Word64` in order to get `Word32` anyways. I just wanted to confirm this was not an oversight.